ykrt/src/compile/jitc_yk/trace_builder.rs

```diff
@@ -1,44 +1,127 @@
 //! The trace builder.
-//!
-//! Given a mapped trace and an AOT module, assembles an in-memory Yk IR trace by copying blocks
-//! from the AOT IR. The output of this process will be the input to the code generator.
 
-use super::aot_ir::{self, Module};
+use super::aot_ir::{self, IRDisplay, Module};
+use super::jit_ir;
 use crate::trace::TracedAOTBlock;
+use std::collections::HashMap;
 use std::error::Error;
 
+/// Given a mapped trace and an AOT module, assembles an in-memory Yk IR trace by copying blocks
+/// from the AOT IR. The output of this process will be the input to the code generator.
 struct TraceBuilder<'a> {
+    /// The AOR IR.
     aot_mod: &'a Module,
-    jit_mod: Module,
+    /// The JIT IR this struct builds.
+    jit_mod: jit_ir::Module,
+    /// The mapped trace.
     mtrace: &'a Vec<TracedAOTBlock>,
+    // Maps an AOT instruction to a jit instruction via their index-based IDs.
+    local_map: HashMap<aot_ir::InstructionID, jit_ir::InstructionID>,
 }
 
 impl<'a> TraceBuilder<'a> {
-    fn new(aot_mod: &'a Module, mtrace: &'a Vec<TracedAOTBlock>) -> Self {
+    /// Create a trace builder.
+    ///
+    /// Arguments:
+    ///  - `trace_name`: The eventual symbol name for the JITted code.
+    ///  - `aot_mod`: The AOT IR module that the trace flows through.
+    ///  - `mtrace`: The mapped trace.
+    fn new(trace_name: String, aot_mod: &'a Module, mtrace: &'a Vec<TracedAOTBlock>) -> Self {
         Self {
             aot_mod,
             mtrace,
-            jit_mod: Module::default(),
+            jit_mod: jit_ir::Module::new(trace_name),
+            local_map: HashMap::new(),
         }
     }
 
-    fn lookup_aot_block(&self, tb: &TracedAOTBlock) -> Option<&aot_ir::Block> {
+    // Given a mapped block, find the AOT block ID, or return `None` if it is unmapped.
+    fn lookup_aot_block(&self, tb: &TracedAOTBlock) -> Option<aot_ir::BlockID> {
         match tb {
             TracedAOTBlock::Mapped { func_name, bb } => {
                 let func_name = func_name.to_str().unwrap(); // safe: func names are valid UTF-8.
-                let func = self.aot_mod.func_by_name(func_name)?;
-                func.block(*bb)
+                let func = self.aot_mod.func_index(func_name)?;
+                Some(aot_ir::BlockID::new(func, *bb))
             }
             TracedAOTBlock::Unmappable { .. } => None,
         }
     }
 
-    fn build(self) -> Result<Module, Box<dyn Error>> {
+    /// Create the prolog of the trace.
+    fn create_trace_header(&mut self, blk: &aot_ir::Block) {
+        // Find trace input variables and emit `LoadArg` instructions for them.
+        let mut last_store = None;
+        let mut trace_input = None;
+        let mut input = Vec::new();
+        for inst in blk.instrs.iter().rev() {
+            if inst.is_control_point(self.aot_mod) {
+                trace_input = Some(inst.get_operand(3));
```

Comment:
How do we know what operand `3` is?

Reply:
We should define a named constant.

Reply:
Fixed in f4d906419449c702878e47a163a0ed1fc387b34f